presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/StateSerializer.java

```diff
@@ -0,0 +1,282 @@
+                        "Inconsistent strategy");
+            }
+            ((FixedHistogramMLEStateStrategy) target).mergeWith((FixedHistogramMLEStateStrategy) source);
+            return;
+        }
+
+        if (target instanceof FixedHistogramJacknifeStateStrategy) {
+            if (!(source instanceof FixedHistogramJacknifeStateStrategy)) {
+                throw new PrestoException(
+                        INVALID_FUNCTION_ARGUMENT,
+                        "Inconsistent strategy");
+            }
+            ((FixedHistogramJacknifeStateStrategy) target).mergeWith((FixedHistogramJacknifeStateStrategy) source);
+            return;
+        }
+
+        throw new InvalidParameterException("unknown strategy combination");
+    }
+
+    @Override
+    public Type getSerializedType()
+    {
+        return VARBINARY;
+    }
+
+    @Override
+    public void serialize(State state, BlockBuilder out)
+    {
+        final int requiredBytes =
+                SizeOf.SIZE_OF_INT + // Method
+                        (state.getStrategy() == null ? 0 : state.getStrategy().getRequiredBytesForSerialization());
+
+        final StateStrategy strategy = state.getStrategy();
+
+        SliceOutput sliceOut = Slices.allocate(requiredBytes).getOutput();
+
+        if (strategy == null) {
+            sliceOut.appendInt(0);
+        }
+        else if (strategy instanceof UnweightedReservoirSampleStateStrategy) {
+            sliceOut.appendInt(1);
+        }
+        else if (strategy instanceof WeightedReservoirSampleStateStrategy) {
+            sliceOut.appendInt(2);
+        }
+        else if (strategy instanceof FixedHistogramMLEStateStrategy) {
+            sliceOut.appendInt(3);
+        }
+        else if (strategy instanceof FixedHistogramJacknifeStateStrategy) {
+            sliceOut.appendInt(4);
+        }
+        else {
+            throw new InvalidParameterException("unknown method in serialize");
+        }
+
+        if (strategy != null) {
+            strategy.serialize(sliceOut);
+        }
+
+        VARBINARY.writeSlice(out, sliceOut.getUnderlyingSlice());
+    }
+
+    @Override
+    public void deserialize(
+            Block block,
+            int index,
+            State state)
+    {
+        StateSerializer.deserialize(
+                VARBINARY.getSlice(block, index).getInput(),
+                state);
+    }
+
+    public static void deserialize(
+            SliceInput input,
+            State state)
+    {
+        final StateStrategy strategy =
+                StateSerializer.deserializeStrategy(input);
+        if (strategy == null && state.getStrategy() != null) {
+            throw new PrestoException(
+                    CONSTRAINT_VIOLATION,
+                    "strategy is not null for null method");
+        }
+        if (strategy != null) {
+            state.setStrategy(strategy);
+        }
+    }
+
+    public static StateStrategy deserializeStrategy(SliceInput input)
```

Comment:
- make this private
- use switch statement
- use verify instead of throw

```
    private static StateStrategy deserializeStrategy(SliceInput input)
    {
        int method = input.readInt();
        switch(method) {
            case 0:
                return null;
            case 1:
                return new UnweightedReservoirSampleStateStrategy(input);
            case 2:
                return new WeightedReservoirSampleStateStrategy(input);
            case 3:
                return new FixedHistogramMLEStateStrategy(input);
            case 4:
                return new FixedHistogramJacknifeStateStrategy(input);
            default:
                verify(false, "Unexpected entropy method: " + method);
                return null;
        }
    }
```